// @alch/alchemy-web3/dist/esm/web3-adapter/alchemyContext.js
import SturdyWebSocket from "sturdy-websocket";
import { w3cwebsocket } from "websocket";
import { makeJsonRpcPayloadFactory, makeJsonRpcSenders, } from "../util/jsonRpc";
import { VERSION } from "../version";
import { makeJsonRpcHttpSender } from "./alchemySendHttp";
import { makeWebSocketSender } from "./alchemySendWebSocket";
import { makeAlchemyHttpProvider } from "./httpProvider";
import { makeJsonRpcPayloadSender } from "./sendJsonRpcPayload";
import { makeRestPayloadSender } from "./sendRestPayload";
import { AlchemyWebSocketProvider } from "./webSocketProvider";
var NODE_MAX_WS_FRAME_SIZE = 100 * 1024 * 1024; // 100 MB
export function makeAlchemyContext(url, config) {
    var makeJsonRpcPayload = makeJsonRpcPayloadFactory();
    var restSender = makeRestPayloadSender({
        config: config,
        url: url,
    });
    if (/^https?:\/\//.test(url)) {
        var alchemySendJsonrRpc = makeJsonRpcHttpSender(url);
        var _a = makeJsonRpcPayloadSender(alchemySendJsonrRpc, config), sendJsonRpcPayload = _a.sendJsonRpcPayload, setWriteProvider = _a.setWriteProvider;
        var jsonRpcSenders = makeJsonRpcSenders(sendJsonRpcPayload, makeJsonRpcPayload);
        var provider = makeAlchemyHttpProvider(sendJsonRpcPayload);
        return { provider: provider, jsonRpcSenders: jsonRpcSenders, restSender: restSender, setWriteProvider: setWriteProvider };
    }
    else if (/^wss?:\/\//.test(url)) {
        var protocol = isAlchemyUrl(url) ? "alchemy-web3-" + VERSION : undefined;
        var ws = new SturdyWebSocket(url, protocol, {
            wsConstructor: getWebSocketConstructor(),
        });
        var alchemySend = makeWebSocketSender(ws);
        var _b = makeJsonRpcPayloadSender(alchemySend, config), sendJsonRpcPayload = _b.sendJsonRpcPayload, setWriteProvider = _b.setWriteProvider;
        var jsonRpcSenders = makeJsonRpcSenders(sendJsonRpcPayload, makeJsonRpcPayload);
        var provider = new AlchemyWebSocketProvider(ws, sendJsonRpcPayload, jsonRpcSenders);
        return { provider: provider, jsonRpcSenders: jsonRpcSenders, restSender: restSender, setWriteProvider: setWriteProvider };
    }
    else {
        throw new Error("Alchemy URL protocol must be one of http, https, ws, or wss. Recieved: " + url);
    }
}
function getWebSocketConstructor() {
    return isNodeEnvironment()
        ? function (url, protocols) {
            return new w3cwebsocket(url, protocols, undefined, undefined, undefined, {
                maxReceivedMessageSize: NODE_MAX_WS_FRAME_SIZE,
                maxReceivedFrameSize: NODE_MAX_WS_FRAME_SIZE,
            });
        }
        : WebSocket;
}
function isNodeEnvironment() {
    return (typeof process !== "undefined" &&
        process != null &&
        process.versions != null &&
        process.versions.node != null);
}
function isAlchemyUrl(url) {
    return url.indexOf("alchemyapi.io") >= 0;
}
