// @alch/alchemy-web3/dist/cjs/web3-adapter/alchemyContext.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeAlchemyContext = void 0;
var tslib_1 = require("tslib");
var sturdy_websocket_1 = tslib_1.__importDefault(require("sturdy-websocket"));
var websocket_1 = require("websocket");
var jsonRpc_1 = require("../util/jsonRpc");
var version_1 = require("../version");
var alchemySendHttp_1 = require("./alchemySendHttp");
var alchemySendWebSocket_1 = require("./alchemySendWebSocket");
var httpProvider_1 = require("./httpProvider");
var sendJsonRpcPayload_1 = require("./sendJsonRpcPayload");
var sendRestPayload_1 = require("./sendRestPayload");
var webSocketProvider_1 = require("./webSocketProvider");
var NODE_MAX_WS_FRAME_SIZE = 100 * 1024 * 1024; // 100 MB
function makeAlchemyContext(url, config) {
    var makeJsonRpcPayload = jsonRpc_1.makeJsonRpcPayloadFactory();
    var restSender = sendRestPayload_1.makeRestPayloadSender({
        config: config,
        url: url,
    });
    if (/^https?:\/\//.test(url)) {
        var alchemySendJsonrRpc = alchemySendHttp_1.makeJsonRpcHttpSender(url);
        var _a = sendJsonRpcPayload_1.makeJsonRpcPayloadSender(alchemySendJsonrRpc, config), sendJsonRpcPayload = _a.sendJsonRpcPayload, setWriteProvider = _a.setWriteProvider;
        var jsonRpcSenders = jsonRpc_1.makeJsonRpcSenders(sendJsonRpcPayload, makeJsonRpcPayload);
        var provider = httpProvider_1.makeAlchemyHttpProvider(sendJsonRpcPayload);
        return { provider: provider, jsonRpcSenders: jsonRpcSenders, restSender: restSender, setWriteProvider: setWriteProvider };
    }
    else if (/^wss?:\/\//.test(url)) {
        var protocol = isAlchemyUrl(url) ? "alchemy-web3-" + version_1.VERSION : undefined;
        var ws = new sturdy_websocket_1.default(url, protocol, {
            wsConstructor: getWebSocketConstructor(),
        });
        var alchemySend = alchemySendWebSocket_1.makeWebSocketSender(ws);
        var _b = sendJsonRpcPayload_1.makeJsonRpcPayloadSender(alchemySend, config), sendJsonRpcPayload = _b.sendJsonRpcPayload, setWriteProvider = _b.setWriteProvider;
        var jsonRpcSenders = jsonRpc_1.makeJsonRpcSenders(sendJsonRpcPayload, makeJsonRpcPayload);
        var provider = new webSocketProvider_1.AlchemyWebSocketProvider(ws, sendJsonRpcPayload, jsonRpcSenders);
        return { provider: provider, jsonRpcSenders: jsonRpcSenders, restSender: restSender, setWriteProvider: setWriteProvider };
    }
    else {
        throw new Error("Alchemy URL protocol must be one of http, https, ws, or wss. Recieved: " + url);
    }
}
exports.makeAlchemyContext = makeAlchemyContext;
function getWebSocketConstructor() {
    return isNodeEnvironment()
        ? function (url, protocols) {
            return new websocket_1.w3cwebsocket(url, protocols, undefined, undefined, undefined, {
                maxReceivedMessageSize: NODE_MAX_WS_FRAME_SIZE,
                maxReceivedFrameSize: NODE_MAX_WS_FRAME_SIZE,
            });
        }
        : WebSocket;
}
function isNodeEnvironment() {
    return (typeof process !== "undefined" &&
        process != null &&
        process.versions != null &&
        process.versions.node != null);
}
function isAlchemyUrl(url) {
    return url.indexOf("alchemyapi.io") >= 0;
}
