// @alch/alchemy-web3/dist/cjs/subscriptions/subscriptionBackfill.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.dedupeLogs = exports.dedupeNewHeads = exports.makeBackfiller = void 0;
var tslib_1 = require("tslib");
var hex_1 = require("../util/hex");
var promises_1 = require("../util/promises");
/**
 * The maximum number of blocks to backfill. If more than this many blocks have
 * been missed, then we'll sadly miss data, but we want to make sure we don't
 * end up requesting thousands of blocks if somebody left their laptop closed
 * for a week.
 */
var MAX_BACKFILL_BLOCKS = 120;
function makeBackfiller(jsonRpcSenders) {
    return { getNewHeadsBackfill: getNewHeadsBackfill, getLogsBackfill: getLogsBackfill };
    function getNewHeadsBackfill(isCancelled, previousHeads, fromBlockNumber) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var toBlockNumber, lastSeenBlockNumber, minBlockNumber, reorgHeads, intermediateHeads;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        promises_1.throwIfCancelled(isCancelled);
                        return [4 /*yield*/, getBlockNumber()];
                    case 1:
                        toBlockNumber = _a.sent();
                        promises_1.throwIfCancelled(isCancelled);
                        if (previousHeads.length === 0) {
                            return [2 /*return*/, getHeadEventsInRange(Math.max(fromBlockNumber, toBlockNumber - MAX_BACKFILL_BLOCKS) + 1, toBlockNumber + 1)];
                        }
                        lastSeenBlockNumber = hex_1.fromHex(previousHeads[previousHeads.length - 1].number);
                        minBlockNumber = Math.max(0, lastSeenBlockNumber - MAX_BACKFILL_BLOCKS);
                        if (lastSeenBlockNumber < minBlockNumber) {
                            return [2 /*return*/, getHeadEventsInRange(minBlockNumber, toBlockNumber + 1)];
                        }
                        return [4 /*yield*/, getReorgHeads(isCancelled, previousHeads)];
                    case 2:
                        reorgHeads = _a.sent();
                        promises_1.throwIfCancelled(isCancelled);
                        return [4 /*yield*/, getHeadEventsInRange(lastSeenBlockNumber + 1, toBlockNumber + 1)];
                    case 3:
                        intermediateHeads = _a.sent();
                        promises_1.throwIfCancelled(isCancelled);
                        return [2 /*return*/, tslib_1.__spreadArray(tslib_1.__spreadArray([], tslib_1.__read(reorgHeads)), tslib_1.__read(intermediateHeads))];
                }
            });
        });
    }
    function getReorgHeads(isCancelled, previousHeads) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var result, i, oldEvent, blockHead;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        result = [];
                        i = previousHeads.length - 1;
                        _a.label = 1;
                    case 1:
                        if (!(i >= 0)) return [3 /*break*/, 4];
                        oldEvent = previousHeads[i];
                        return [4 /*yield*/, getBlockByNumber(hex_1.fromHex(oldEvent.number))];
                    case 2:
                        blockHead = _a.sent();
                        promises_1.throwIfCancelled(isCancelled);
                        if (oldEvent.hash === blockHead.hash) {
                            return [3 /*break*/, 4];
                        }
                        result.push(toNewHeadsEvent(blockHead));
                        _a.label = 3;
                    case 3:
                        i--;
                        return [3 /*break*/, 1];
                    case 4: return [2 /*return*/, result.reverse()];
                }
            });
        });
    }
    function getHeadEventsInRange(fromBlockInclusive, toBlockExclusive) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var batchParts, i, heads;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (fromBlockInclusive >= toBlockExclusive) {
                            return [2 /*return*/, []];
                        }
                        batchParts = [];
                        for (i = fromBlockInclusive; i < toBlockExclusive; i++) {
                            batchParts.push({
                                method: "eth_getBlockByNumber",
                                params: [hex_1.toHex(i), false],
                            });
                        }
                        return [4 /*yield*/, jsonRpcSenders.sendBatch(batchParts)];
                    case 1:
                        heads = _a.sent();
                        return [2 /*return*/, heads.map(toNewHeadsEvent)];
                }
            });
        });
    }
    function getBlockByNumber(blockNumber) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            return tslib_1.__generator(this, function (_a) {
                return [2 /*return*/, jsonRpcSenders.send("eth_getBlockByNumber", [
                        hex_1.toHex(blockNumber),
                        false,
                    ])];
            });
        });
    }
    function getLogsBackfill(isCancelled, filter, previousLogs, fromBlockNumber) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var toBlockNumber, lastSeenBlockNumber, minBlockNumber, commonAncestorNumber, removedLogs, addedLogs;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        promises_1.throwIfCancelled(isCancelled);
                        return [4 /*yield*/, getBlockNumber()];
                    case 1:
                        toBlockNumber = _a.sent();
                        promises_1.throwIfCancelled(isCancelled);
                        if (previousLogs.length === 0) {
                            return [2 /*return*/, getLogsInRange(filter, Math.max(fromBlockNumber, toBlockNumber - MAX_BACKFILL_BLOCKS) + 1, toBlockNumber + 1)];
                        }
                        lastSeenBlockNumber = hex_1.fromHex(previousLogs[previousLogs.length - 1].blockNumber);
                        minBlockNumber = Math.max(0, lastSeenBlockNumber - MAX_BACKFILL_BLOCKS);
                        if (lastSeenBlockNumber < minBlockNumber) {
                            return [2 /*return*/, getLogsInRange(filter, minBlockNumber, toBlockNumber + 1)];
                        }
                        return [4 /*yield*/, getCommonAncestorNumber(isCancelled, previousLogs)];
                    case 2:
                        commonAncestorNumber = _a.sent();
                        promises_1.throwIfCancelled(isCancelled);
                        removedLogs = previousLogs
                            .filter(function (log) { return hex_1.fromHex(log.blockNumber) > commonAncestorNumber; })
                            .map(function (log) { return (tslib_1.__assign(tslib_1.__assign({}, log), { removed: true })); });
                        return [4 /*yield*/, getLogsInRange(filter, commonAncestorNumber + 1, toBlockNumber + 1)];
                    case 3:
                        addedLogs = _a.sent();
                        promises_1.throwIfCancelled(isCancelled);
                        return [2 /*return*/, tslib_1.__spreadArray(tslib_1.__spreadArray([], tslib_1.__read(removedLogs)), tslib_1.__read(addedLogs))];
                }
            });
        });
    }
    function getCommonAncestorNumber(isCancelled, previousLogs) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var i, _a, blockHash, blockNumber, hash;
            return tslib_1.__generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        i = previousLogs.length - 1;
                        _b.label = 1;
                    case 1:
                        if (!(i >= 0)) return [3 /*break*/, 4];
                        _a = previousLogs[i], blockHash = _a.blockHash, blockNumber = _a.blockNumber;
                        return [4 /*yield*/, getBlockByNumber(hex_1.fromHex(blockNumber))];
                    case 2:
                        hash = (_b.sent()).hash;
                        promises_1.throwIfCancelled(isCancelled);
                        if (blockHash === hash) {
                            return [2 /*return*/, hex_1.fromHex(blockNumber)];
                        }
                        _b.label = 3;
                    case 3:
                        i--;
                        return [3 /*break*/, 1];
                    case 4: return [2 /*return*/, Number.NEGATIVE_INFINITY];
                }
            });
        });
    }
    function getLogsInRange(filter, fromBlockInclusive, toBlockExclusive) {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var rangeFilter;
            return tslib_1.__generator(this, function (_a) {
                if (fromBlockInclusive >= toBlockExclusive) {
                    return [2 /*return*/, []];
                }
                rangeFilter = tslib_1.__assign(tslib_1.__assign({}, filter), { fromBlock: hex_1.toHex(fromBlockInclusive), toBlock: hex_1.toHex(toBlockExclusive - 1) });
                return [2 /*return*/, jsonRpcSenders.send("eth_getLogs", [rangeFilter])];
            });
        });
    }
    function getBlockNumber() {
        return tslib_1.__awaiter(this, void 0, void 0, function () {
            var blockNumberHex;
            return tslib_1.__generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, jsonRpcSenders.send("eth_blockNumber")];
                    case 1:
                        blockNumberHex = _a.sent();
                        return [2 /*return*/, hex_1.fromHex(blockNumberHex)];
                }
            });
        });
    }
}
exports.makeBackfiller = makeBackfiller;
function toNewHeadsEvent(head) {
    var result = tslib_1.__assign({}, head);
    delete result.totalDifficulty;
    delete result.transactions;
    delete result.uncles;
    return result;
}
function dedupeNewHeads(events) {
    return dedupe(events, function (event) { return event.hash; });
}
exports.dedupeNewHeads = dedupeNewHeads;
function dedupeLogs(events) {
    return dedupe(events, function (event) { return event.blockHash + "/" + event.logIndex; });
}
exports.dedupeLogs = dedupeLogs;
function dedupe(items, getKey) {
    var keysSeen = new Set();
    var result = [];
    items.forEach(function (item) {
        var key = getKey(item);
        if (!keysSeen.has(key)) {
            keysSeen.add(key);
            result.push(item);
        }
    });
    return result;
}
